ethereumj-core/src/main/java/org/ethereum/casper/core/genesis/CasperStateInit.java

```diff
@@ -0,0 +1,200 @@
+import org.ethereum.casper.config.CasperProperties;
+import org.ethereum.config.SystemProperties;
+import org.ethereum.core.Blockchain;
+import org.ethereum.core.CallTransaction;
+import org.ethereum.core.Genesis;
+import org.ethereum.core.Repository;
+import org.ethereum.core.Transaction;
+import org.ethereum.core.genesis.StateInit;
+import org.ethereum.crypto.ECKey;
+import org.ethereum.db.BlockStore;
+import org.ethereum.util.ByteUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.spongycastle.util.encoders.Hex;
+
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.ethereum.casper.config.net.CasperTestNetConfig.BASE_INTEREST_FACTOR;
+import static org.ethereum.casper.config.net.CasperTestNetConfig.BASE_PENALTY_FACTOR;
+import static org.ethereum.casper.config.net.CasperTestNetConfig.MIN_DEPOSIT_ETH;
+import static org.ethereum.casper.config.net.CasperTestNetConfig.NULL_SIGN_SENDER;
+import static org.ethereum.casper.config.net.CasperTestNetConfig.WITHDRAWAL_DELAY;
+import static org.ethereum.crypto.HashUtil.sha3;
+
+public class CasperStateInit implements StateInit {
+
+    private static final Logger logger = LoggerFactory.getLogger("general");
+
+    private Genesis genesis;
+
+    private Repository repository;
+
+    private Blockchain blockchain;
+
+    private CasperProperties systemProperties;
+
+    private Genesis initGenesis;
+
+    public CasperStateInit(Genesis genesis, Repository repository, Blockchain blockchain,
+                           SystemProperties systemProperties) {
+        this.genesis = genesis;
+        this.repository = repository;
+        this.blockchain = blockchain;
+        this.systemProperties = (CasperProperties) systemProperties;
+        init();
+    }
+
+    @Override
+    public void initDB() {
+        if (blockchain.getBlockByNumber(0) != null) {
+            return;  // Already initialized
+        }
+        logger.info("DB is empty - initializing Casper Genesis");
+        Genesis.populateRepository(repository, genesis);
+
+        BlockStore blockStore = ((org.ethereum.facade.Blockchain)blockchain).getBlockStore();
+
+
+        // Metropolis dummy
+        repository.saveCode(Hex.decode("0000000000000000000000000000000000000010"), Hex.decode("6000355460205260206020f3"));
+        repository.saveCode(Hex.decode("0000000000000000000000000000000000000020"), Hex.decode("6000355460205260206020f3"));
+        repository.commit();
+        genesis.setStateRoot(repository.getRoot());
+
+        blockStore.saveBlock(genesis, genesis.getCumulativeDifficulty(), true);
+        blockchain.setBestBlock(genesis);
+        blockchain.setTotalDifficulty(genesis.getCumulativeDifficulty());
+
+        logger.info("Genesis block loaded");
+    }
+
+    /**
+     * @return key: Casper contract address
+     *       value: Casper state initialization transactions
+     */
+    public Pair<byte[], List<Transaction>> makeInitTxes() {
```

Comment:
I think these Txes is better to keep in `CasperFacade` responsibility scope

Reply:
Yeah, looks like we don't need StateInit abstraction at all, will try to get rid of it.